tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+            }
+
+            let mut output_carry = block_add_assign_returning_carry(
+                self,
+                &mut lhs[num_blocks - 1],
+                &rhs[num_blocks - 1],
+                &carry,
+            );
+
+            return match requested_flag {
+                OutputFlag::None => None,
+                OutputFlag::Overflow => {
+                    let overflowed = self.key.not_equal(&output_carry, &carry);
+                    Some(BooleanBlock::new_unchecked(overflowed))
+                }
+                OutputFlag::Carry => {
+                    output_carry.degree = Degree::new(1);
+                    Some(BooleanBlock::new_unchecked(output_carry))
+                }
+            };
+        }
+
+        panic!(
+            "Invalid combo of message modulus ({}) and carry modulus ({}) \n\
+            This function requires the message modulus >= 2 and carry modulus >= message_modulus \n\
+            I.e. PARAM_MESSAGE_X_CARRY_Y where X >= 1 and Y >= X.",
+            self.key.message_modulus.0, self.key.carry_modulus.0
+        );
     }
 
-    /// This function takes an input slice of shortint ciphertext (aka blocks)
-    /// for which at most one bit of carry is consumed in each block, and
-    /// it does the carry propagation in place.
+    /// Does lhs += (rhs + carry)
     ///
-    /// It returns the output carry of the last block
+    /// acts like the ADC assemby op, except, the flags have to be explicitely requested
+    /// as they incur additional PBSes
     ///
-    /// Used in (among other) 'default' addition:
-    /// - first unchecked_add
-    /// - at this point at most on bit of carry is taken
-    /// - use this function to propagate them in parallel
-    pub(crate) fn propagate_single_carry_parallelized_low_latency(
+    /// - Parameters must have at least 2 bits of message, 2 bits of carry
+    /// - blocks of lhs and rhs must be clean (no carries)
+    /// - lhs and rhs must have the same length
+    pub(crate) fn advanced_add_assign_with_carry_at_least_4_bits(
         &self,
-        blocks: &mut [Ciphertext],
-    ) -> Ciphertext {
-        let generates_or_propagates = self.generate_init_carry_array(blocks);
-        let (input_carries, output_carry) =
-            self.compute_carry_propagation_parallelized_low_latency(generates_or_propagates);
+        lhs: &mut [Ciphertext],
+        rhs: &[Ciphertext],
+        input_carry: Option<&BooleanBlock>,
+        requested_flag: OutputFlag,
+    ) -> Option<BooleanBlock> {
+        // Empty rhs is a specially allowed 'weird' case to
+        // act like a 'propagate single carry' function.
+        // This is not made explicit in the docs as we have a
+        // `propagate_single_carry_parallelized` function which wraps this special case
+        if rhs.is_empty() {
+            // Techinically, CarryFlag is computable, but OverflowFlag is not
+            assert_eq!(requested_flag, OutputFlag::None);
+        } else {
```

Comment:
make rhs optional ? or even (rhs, output_flag) ?

is the special case dangerous somehow ?